opensafely/launch.py

```diff
@@ -0,0 +1,207 @@
+import argparse
+import json
+import os
+import subprocess
+import sys
+import time
+from pathlib import Path
+
+from opensafely import utils
+from opensafely._vendor.jobrunner.cli.local_run import docker_preflight_check
+
+
+DESCRIPTION = "Launch an RStudio Server or Jupyter Lab session in a browser"
+
+
+def add_base_args(parser):
+    parser.add_argument(
+        "--directory",
+        "-d",
+        default=os.getcwd(),
+        type=Path,
+        help="Directory to run the RStudio Server session in (default is current dir)",
+    )
+    parser.add_argument(
+        "--name", help="Name of docker container (defaults to use directory name)"
+    )
+    parser.add_argument(
+        "--port",
+        "-p",
+        default=None,
+        help="Port to run on (random by default)",
+    )
+    parser.add_argument(
+        "--no-browser",
+        "-n",
+        default=False,
+        action="store_true",
+        help="Do not attempt to open a browser",
+    )
+
+
+def add_arguments(parser):
+    parser.add_argument(
+        "tool",
+        help="Tool to launch e.g. rstudio",
```

Comment:
This could have a `choices` arg? Then it shouldn't need to raise the `ArgumentTypeError` because it would only accept rstudio/jupyter options

Reply:
Actually, now I see that we still need that check in `main` because the deprecated commands are calling it directly. Adding choices here would still give you a nicer help text that tells you what the options are though

Reply:
Good idea

Reply:
Ah, I remember why I didn't to this.

I don't want to have to pre-enumerate all the versions, i.e. 
```choices=["jupyter", "jupyter:v1", "jupyter:v2", "rstudio", "rstudio:v1", "rstudio:v2", ...]
```

I'd like to to work with new versions of image w/o updates.

My custom validation effectively checks that the name before the `:` is valid. You could make argparse do this, but I think it ends up more complex.